File: 16_Bug/javascript/bug.js
// BUG
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    return new Promise(function (resolve) {
                       const input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_element.addEventListener("keydown",
                           function (event) {
                                      if (event.keyCode === 13) {
                                          const input_str = input_element.value;
                                          document.getElementById("output").removeChild(input_element);
                                          print(input_str);
                                          print("\n");
                                          resolve(input_str);
                                      }
                                  });
                       });
}

function tab(space)
{
    let str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

function waitNSeconds(n) {
    return new Promise(resolve => setTimeout(resolve, n*1000));
}

function scrollToBottom() {
    window.scrollTo(0, document.body.scrollHeight);
}

function draw_head()
{
    print("        HHHHHHH\n");
    print("        H     H\n");
    print("        H O O H\n");
    print("        H     H\n");
    print("        H  V  H\n");
    print("        HHHHHHH\n");
}

function drawFeelers(feelerCount, character) {
    for (let z = 1; z <= 4; z++) {
        print(tab(10));
        for (let x = 1; x <= feelerCount; x++) {
            print(character + " ");
        }
        print("\n");
    }
}

function drawNeck() {
    for (let z = 1; z <= 2; z++)
        print("          N N\n");
}

function drawBody(computerTailCount) {
    print("     BBBBBBBBBBBB\n");
    for (let z = 1; z <= 2; z++)
        print("     B          B\n");
    if (computerTailCount === 1)
        print("TTTTTB          B\n");
    print("     BBBBBBBBBBBB\n");
}

function drawFeet(computerFeetCount) {
    for (let z = 1; z <= 2; z++) {
        print(tab(5));
        for (let x = 1; x <= computerFeetCount; x++)
            print(" L");
        print("\n");
    }
}

function drawBug(playerFeelerCount, playerHeadCount, playerNeckCount, playerBodyCount, playerTailCount, playerFeetCount, feelerCharacter) {
    if (playerFeelerCount !== 0) {
        drawFeelers(playerFeelerCount, feelerCharacter);
    }
    if (playerHeadCount !== 0)
        draw_head();
    if (playerNeckCount !== 0) {
        drawNeck();
    }
    if (playerBodyCount !== 0) {
        drawBody(playerTailCount)
    }
    if (playerFeetCount !== 0) {
        drawFeet(playerFeetCount);
    }
    for (let z = 1; z <= 4; z++)
        print("\n");
}

// Main program
async function main()
{
    print(tab(34) + "BUG\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    let playerFeelerCount = 0;
    let playerHeadCount = 0;
    let playerNeckCount = 0;
    let playerBodyCount = 0;
    let playerFeetCount = 0;
    let playerTailCount = 0;

    let computerFeelerCount = 0;
    let computerHeadCount = 0;
    let computerNeckCount = 0;
    let computerBodyCount = 0;
    let computerTailCount = 0;
    let computerFeetCount = 0;

    print("THE GAME BUG\n");
    print("I HOPE YOU ENJOY THIS GAME.\n");
    print("\n");
    print("DO YOU WANT INSTRUCTIONS");
    const instructionsRequired = await input();
    if (instructionsRequired.toUpperCase() !== "NO") {
        print("THE OBJECT OF BUG IS TO FINISH YOUR BUG BEFORE I FINISH\n");
        print("MINE. EACH NUMBER STANDS FOR A PART OF THE BUG BODY.\n");
        print("I WILL ROLL THE DIE FOR YOU, TELL YOU WHAT I ROLLED FOR YOU\n");
        print("WHAT THE NUMBER STANDS FOR, AND IF YOU CAN GET THE PART.\n");
        print("IF YOU CAN GET THE PART I WILL GIVE IT TO YOU.\n");
        print("THE SAME WILL HAPPEN ON MY TURN.\n");
        print("IF THERE IS A CHANGE IN EITHER BUG I WILL GIVE YOU THE\n");
        print("OPTION OF SEEING THE PICTURES OF THE BUGS.\n");
        print("THE NUMBERS STAND FOR PARTS AS FOLLOWS:\n");
        print("NUMBER\tPART\tNUMBER OF PART NEEDED\n");
        print("1\tBODY\t1\n");
        print("2\tNECK\t1\n");
        print("3\tHEAD\t1\n");
        print("4\tFEELERS\t2\n");
        print("5\tTAIL\t1\n");
        print("6\tLEGS\t6\n");
        print("\n");
        print("\n");
    }

    let gameInProgress = true;
    while (gameInProgress) {
        let dieRoll = Math.floor(6 * Math.random() + 1);
        let partFound = false;
        print("YOU ROLLED A " + dieRoll + "\n");
        switch (dieRoll) {
            case 1:
                print("1=BODY\n");
                if (playerBodyCount === 0) {
                    print("YOU NOW HAVE A BODY.\n");
                    playerBodyCount = 1;
                    partFound = true;
                } else {
                    print("YOU DO NOT NEED A BODY.\n");
                }
                break;
            case 2:
                print("2=NECK\n");
                if (playerNeckCount === 0) {
                    if (playerBodyCount === 0) {
                        print("YOU DO NOT HAVE A BODY.\n");
                    } else {
                        print("YOU NOW HAVE A NECK.\n");
                        playerNeckCount = 1;
                        partFound = true;
                    }
                } else {
                    print("YOU DO NOT NEED A NECK.\n");
                }
                break;
            case 3:
                print("3=HEAD\n");
                if (playerNeckCount === 0) {
                    print("YOU DO NOT HAVE A NECK.\n");
                } else if (playerHeadCount === 0) {
                    print("YOU NEEDED A HEAD.\n");
                    playerHeadCount = 1;
                    partFound = true;
                } else {
                    print("YOU HAVE A HEAD.\n");
                }
                break;
            case 4:
                print("4=FEELERS\n");
                if (playerHeadCount === 0) {
                    print("YOU DO NOT HAVE A HEAD.\n");
                } else if (playerFeelerCount === 2) {
                    print("YOU HAVE TWO FEELERS ALREADY.\n");
                } else {
                    print("I NOW GIVE YOU A FEELER.\n");
                    playerFeelerCount ++;
                    partFound = true;
                }
                break;
            case 5:
                print("5=TAIL\n");
                if (playerBodyCount === 0) {
                    print("YOU DO NOT HAVE A BODY.\n");
                } else if (playerTailCount === 1) {
                    print("YOU ALREADY HAVE A TAIL.\n");
                } else {
                    print("I NOW GIVE YOU A TAIL.\n");
                    playerTailCount++;
                    partFound = true;
                }
                break;
            case 6:
                print("6=LEG\n");
                if (playerFeetCount === 6) {
                    print("YOU HAVE 6 FEET ALREADY.\n");
                } else if (playerBodyCount === 0) {
                    print("YOU DO NOT HAVE A BODY.\n");
                } else {
                    playerFeetCount++;
                    partFound = true;
                    print("YOU NOW HAVE " + playerFeetCount + " LEGS.\n");
                }
                break;
        }
        dieRoll = Math.floor(6 * Math.random() + 1) ;
        print("\n");
        scrollToBottom();
        await waitNSeconds(1);

        print("I ROLLED A " + dieRoll + "\n");
        switch (dieRoll) {
            case 1:
                print("1=BODY\n");
                if (computerBodyCount === 1) {
                    print("I DO NOT NEED A BODY.\n");
                } else {
                    print("I NOW HAVE A BODY.\n");
                    partFound = true;
                    computerBodyCount = 1;
                }
                break;
            case 2:
                print("2=NECK\n");
                if (computerNeckCount === 1) {
                    print("I DO NOT NEED A NECK.\n");
                } else if (computerBodyCount === 0) {
                    print("I DO NOT HAVE A BODY.\n");
                } else {
                    print("I NOW HAVE A NECK.\n");
                    computerNeckCount = 1;
                    partFound = true;
                }
                break;
            case 3:
                print("3=HEAD\n");
                if (computerNeckCount === 0) {
                    print("I DO NOT HAVE A NECK.\n");
                } else if (computerHeadCount === 1) {
                    print("I DO NOT NEED A HEAD.\n");
                } else {
                    print("I NEEDED A HEAD.\n");
                    computerHeadCount = 1;
                    partFound = true;
                }
                break;
            case 4:
                print("4=FEELERS\n");
                if (computerHeadCount === 0) {
                    print("I DO NOT HAVE A HEAD.\n");
                } else if (computerFeelerCount === 2) {
                    print("I HAVE 2 FEELERS ALREADY.\n");
                } else {
                    print("I GET A FEELER.\n");
                    computerFeelerCount++;
                    partFound = true;
                }
                break;
            case 5:
                print("5=TAIL\n");
                if (computerBodyCount === 0) {
                    print("I DO NOT HAVE A BODY.\n");
                } else if (computerTailCount === 1) {
                    print("I DO NOT NEED A TAIL.\n");
                } else {
                    print("I NOW HAVE A TAIL.\n");
                    computerTailCount = 1;
                    partFound = true;
                }
                break;
            case 6:
                print("6=LEGS\n");
                if (computerFeetCount === 6) {
                    print("I HAVE 6 FEET.\n");
                } else if (computerBodyCount === 0) {
                    print("I DO NOT HAVE A BODY.\n");
                } else {
                    computerFeetCount++;
                    partFound = true;
                    print("I NOW HAVE " + computerFeetCount + " LEGS.\n");
                }
                break;
        }
        if (playerFeelerCount === 2 && playerTailCount === 1 && playerFeetCount === 6) {
            print("YOUR BUG IS FINISHED.\n");
            gameInProgress = false;
        }
        if (computerFeelerCount === 2 && computerBodyCount === 1 && computerFeetCount === 6) {
            print("MY BUG IS FINISHED.\n");
            gameInProgress = false;
        }
        if (!partFound)
            continue;
        print("DO YOU WANT THE PICTURES");
        const showPictures = await input();
        if (showPictures.toUpperCase() === "NO")
            continue;
        print("*****YOUR BUG*****\n");
        print("\n");
        print("\n");
        drawBug(playerFeelerCount, playerHeadCount, playerNeckCount, playerBodyCount, playerTailCount, playerFeetCount, "A");
        print("*****MY BUG*****\n");
        print("\n");
        print("\n");
        drawBug(computerFeelerCount, computerHeadCount, computerNeckCount, computerBodyCount, computerTailCount, computerFeetCount, "F");
        for (let z = 1; z <= 4; z++)
            print("\n");
    }
    print("I HOPE YOU ENJOYED THE GAME, PLAY IT AGAIN SOON!!\n");
    scrollToBottom();
}

main();

File: 16_Bug/python/bug.py
import random
import time
from dataclasses import dataclass
from typing import Literal


@dataclass
class State:
    is_player: bool
    body: int = 0
    neck: int = 0
    head: int = 0
    feelers: int = 0
    tail: int = 0
    legs: int = 0

    def is_finished(self) -> bool:
        return (
            self.feelers == 2
            and self.tail == 1
            and self.legs == 6
            and self.head == 1
            and self.neck == 1
        )

    def display(self) -> None:
        if self.feelers != 0:
            print_feelers(self.feelers, is_player=self.is_player)
        if self.head != 0:
            print_head()
        if self.neck != 0:
            print_neck()
        if self.body != 0:
            print_body(True) if self.tail == 1 else print_body(False)
        if self.legs != 0:
            print_legs(self.legs)


def print_n_newlines(n: int) -> None:
    for _ in range(n):
        print()


def print_feelers(n_feelers: int, is_player: bool = True) -> None:
    for _ in range(4):
        print(" " * 10, end="")
        for _ in range(n_feelers):
            print("A " if is_player else "F ", end="")
        print()


def print_head() -> None:
    print("        HHHHHHH")
    print("        H     H")
    print("        H O O H")
    print("        H     H")
    print("        H  V  H")
    print("        HHHHHHH")


def print_neck() -> None:
    print("          N N")
    print("          N N")


def print_body(has_tail: bool = False) -> None:
    print("     BBBBBBBBBBBB")
    print("     B          B")
    print("     B          B")
    print("TTTTTB          B") if has_tail else ""
    print("     BBBBBBBBBBBB")


def print_legs(n_legs: int) -> None:
    for _ in range(2):
        print(" " * 5, end="")
        for _ in range(n_legs):
            print(" L", end="")
        print()


def handle_roll(diceroll: Literal[1, 2, 3, 4, 5, 6], state: State) -> bool:
    who = "YOU" if state.is_player else "I"
    changed = False

    print(f"{who} ROLLED A", diceroll)
    if diceroll == 1:
        print("1=BODY")
        if state.body:
            print(f"{who} DO NOT NEED A BODY.")
        else:
            print(f"{who} NOW HAVE A BODY.")
            state.body = 1
            changed = True
    elif diceroll == 2:
        print("2=NECK")
        if state.neck:
            print(f"{who} DO NOT NEED A NECK.")
        elif state.body == 0:
            print(f"{who} DO NOT HAVE A BODY.")
        else:
            print(f"{who} NOW HAVE A NECK.")
            state.neck = 1
            changed = True
    elif diceroll == 3:
        print("3=HEAD")
        if state.neck == 0:
            print(f"{who} DO NOT HAVE A NECK.")
        elif state.head:
            print(f"{who} HAVE A HEAD.")
        else:
            print(f"{who} NEEDED A HEAD.")
            state.head = 1
            changed = True
    elif diceroll == 4:
        print("4=FEELERS")
        if state.head == 0:
            print(f"{who} DO NOT HAVE A HEAD.")
        elif state.feelers == 2:
            print(f"{who} HAVE TWO FEELERS ALREADY.")
        else:
            if state.is_player:
                print("I NOW GIVE YOU A FEELER.")
            else:
                print(f"{who} GET A FEELER.")
            state.feelers += 1
            changed = True
    elif diceroll == 5:
        print("5=TAIL")
        if state.body == 0:
            print(f"{who} DO NOT HAVE A BODY.")
        elif state.tail:
            print(f"{who} ALREADY HAVE A TAIL.")
        else:
            if state.is_player:
                print("I NOW GIVE YOU A TAIL.")
            else:
                print(f"{who} NOW HAVE A TAIL.")
            state.tail = 1
            changed = True
    elif diceroll == 6:
        print("6=LEG")
        if state.legs == 6:
            print(f"{who} HAVE 6 FEET ALREADY.")
        elif state.body == 0:
            print(f"{who} DO NOT HAVE A BODY.")
        else:
            state.legs += 1
            changed = True
            print(f"{who} NOW HAVE {state.legs} LEGS")
    return changed


def main() -> None:
    print(" " * 34 + "BUG")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print_n_newlines(3)

    print("THE GAME BUG")
    print("I HOPE YOU ENJOY THIS GAME.")
    print()
    want_instructions = input("DO YOU WANT INSTRUCTIONS? ")
    if want_instructions != "NO":
        print("THE OBJECT OF BUG IS TO FINISH YOUR BUG BEFORE I FINISH")
        print("MINE. EACH NUMBER STANDS FOR A PART OF THE BUG BODY.")
        print("I WILL ROLL THE DIE FOR YOU, TELL YOU WHAT I ROLLED FOR YOU")
        print("WHAT THE NUMBER STANDS FOR, AND IF YOU CAN GET THE PART.")
        print("IF YOU CAN GET THE PART I WILL GIVE IT TO YOU.")
        print("THE SAME WILL HAPPEN ON MY TURN.")
        print("IF THERE IS A CHANGE IN EITHER BUG I WILL GIVE YOU THE")
        print("OPTION OF SEEING THE PICTURES OF THE BUGS.")
        print("THE NUMBERS STAND FOR PARTS AS FOLLOWS:")
        table = [
            ["NUMBER", "PART", "NUMBER OF PART NEEDED"],
            ["1", "BODY", "1"],
            ["2", "NECK", "1"],
            ["3", "HEAD", "1"],
            ["4", "FEELERS", "2"],
            ["5", "TAIL", "1"],
            ["6", "LEGS", "6"],
        ]
        for row in table:
            print(f"{row[0]:<16}{row[1]:<16}{row[2]:<20}")
        print_n_newlines(2)

    player = State(is_player=True)
    opponent = State(is_player=False)
    bugs_finished = 0

    while bugs_finished <= 0:
        diceroll = random.randint(1, 6)
        print()
        changed = handle_roll(diceroll, player)  # type: ignore

        diceroll = random.randint(1, 6)
        print()
        time.sleep(2)

        changed_op = handle_roll(diceroll, opponent)  # type: ignore

        changed = changed or changed_op

        if player.is_finished():
            print("YOUR BUG IS FINISHED.")
            bugs_finished += 1
        if opponent.is_finished():
            print("MY BUG IS FINISHED.")
            bugs_finished += 1
        if not changed:
            continue
        want_pictures = input("DO YOU WANT THE PICTURES? ")
        if want_pictures != "NO":
            print("*****YOUR BUG*****")
            print_n_newlines(2)
            player.display()
            print_n_newlines(4)
            print("*****MY BUG*****")
            print_n_newlines(3)
            opponent.display()

            if bugs_finished != 0:
                break

    print("I HOPE YOU ENJOYED THE GAME, PLAY IT AGAIN SOON!!")


if __name__ == "__main__":
    main()
